sharedbuffers/mapped_struct.py

```diff
@@ -342,6 +342,194 @@ def unpack_from(cls, buf, offs, idmap = None):
         key, values = mapped_list.unpack_from(buf, offs, idmap)
         return cls(zip(key, values))
 
+
+def _upsize(x):
+    "Round up X to the next power of 2"
+    for p2 in (1, 2, 4, 8, 16, 32, 64):
+        x |= x >> p2
+    return x + 1
+
+def _add_kv(indexes, kvpairs, key, val):
+    l2 = len(indexes) / 2
+    hval = hash(key)
+    index = hval % l2
+    nprobe = 1
+
+    while True:
+        i2 = index + index
+        if indexes[i2] == -1:
+            indexes[i2] = len(kvpairs)
+            indexes[i2 + 1] = hval
+            kvpairs.extend((key, val))
+            return
+
+        index = (index + nprobe) % l2
+        nprobe += 1
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(
+        index_table=tuple,
+        kvpairs=proxied_list,
+        mask=int,
+        nelems=int)
+
+    def __init__(self, mask, nelems, index_table, kvpairs):
+        self.mask = mask
+        self.nelems = nelems
+        self.index_table = index_table
+        self.kvpairs = kvpairs
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        """
+        The binary format of a proxied dictionary is as follows:
+        | size           | type            | attribute name
+        |  8             | long            | mask
+        |  8             | long            | nelems
+        | $mask + 1 * 16 | array of long's | index_table
+        | $nelems        | proxied_list    | kvpairs
+
+        In other words, there are 2 tables: an index vector and a key/value
+        array. The index vector contains the positions in the second array,
+        and their hash values, to speed up comparisons. Unused buckets have
+        a position of -1. The key/value array is packed into a proxied list.
+        """
+
+        xlen = len(obj)
+        size = _upsize(xlen + xlen / 2)   # Aim for a load factor of ~50%
+        indexes = [-1, 0] * size
+        kvpairs = []
```

Comment:
There's no need to keep a list of kvpairs.

You know the size of the index beforehand (in any case, with the hash table or the sorted table `_search_hkey` uses). You can leave enough room in the destination buffer for the index and pack the key-value pairs as you add the items to the index. Then you sort the index in-place and you're done.